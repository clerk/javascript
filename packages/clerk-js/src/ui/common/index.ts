export * from './authPropHelpers';
export * from './constants';
export * from './forms';
export * from './redirects';
export * from './verification';
export * from './withRedirectToHome';
export * from './SSOCallback';
export * from './EmailLinkVerify';
export * from './EmailLinkStatusCard';
export * from './Wizard';
export * from './PrintableComponent';
export * from './withOrganizationsEnabledGuard';
export * from './QRCode';
