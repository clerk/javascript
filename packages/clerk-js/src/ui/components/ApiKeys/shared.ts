import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
