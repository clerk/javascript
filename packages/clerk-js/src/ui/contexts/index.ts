export * from './CoreClerkContext';
export * from './CoreUserContext';
export * from './CoreOrganizationContext';
export * from './ClerkUIComponentsContext';
export * from './EnvironmentContext';
export * from './OptionsContext';
export * from './CoreSessionContext';
export * from './CoreClientContext';
