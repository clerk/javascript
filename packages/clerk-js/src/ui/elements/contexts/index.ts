export * from './CardStateContext';
export * from './FlowMetadataContext';
