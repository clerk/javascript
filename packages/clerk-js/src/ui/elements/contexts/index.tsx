import { createContextAndHook } from '@clerk/shared/react';
import type { ClerkAPIError, ClerkRuntimeError } from '@clerk/types';
import { FloatingTree, useFloatingParentNodeId } from '@floating-ui/react';
import React from 'react';

import { useLocalizations } from '../../customizables';
import { useSafeState } from '../../hooks';

type Status = 'idle' | 'loading' | 'error';
type Metadata = string | undefined;
type State = { status: Status; metadata: Metadata; error: string | undefined };
type CardStateCtxValue = {
  state: State;
  setState: React.Dispatch<React.SetStateAction<State>>;
};

const [CardStateCtx, _useCardState] = createContextAndHook<CardStateCtxValue>('CardState');

export const CardStateProvider = (props: React.PropsWithChildren<any>) => {
  const { translateError } = useLocalizations();

  const [state, setState] = useSafeState<State>({
    status: 'idle',
    metadata: undefined,
    error: translateError(window?.Clerk?.__internal_last_error || undefined),
  });

  const value = React.useMemo(() => ({ value: { state, setState } }), [state, setState]);
  return <CardStateCtx.Provider value={value}>{props.children}</CardStateCtx.Provider>;
};

export const useCardState = () => {
  const { state, setState } = _useCardState();
  const { translateError } = useLocalizations();

  const setIdle = (metadata?: Metadata) => setState(s => ({ ...s, status: 'idle', metadata }));
  const setError = (metadata: ClerkRuntimeError | ClerkAPIError | Metadata | string) =>
    setState(s => ({ ...s, error: translateError(metadata) }));
  const setLoading = (metadata?: Metadata) => setState(s => ({ ...s, status: 'loading', metadata }));
  const runAsync = async <T = unknown,>(cb: Promise<T> | (() => Promise<T>), metadata?: Metadata) => {
    setLoading(metadata);
    return (typeof cb === 'function' ? cb() : cb)
      .then(res => {
        return res;
      })
      .finally(() => setIdle(metadata));
  };

  return {
    setIdle,
    setError,
    setLoading,
    runAsync,
    loadingMetadata: state.status === 'loading' ? state.metadata : undefined,
    error: state.error ? state.error : undefined,
    isLoading: state.status === 'loading',
    isIdle: state.status === 'idle',
    state,
  };
};

export const withCardStateProvider = <T,>(Component: React.ComponentType<T>) => {
  return (props: T) => {
    return (
      <CardStateProvider>
        {/* @ts-expect-error */}
        <Component {...props} />
      </CardStateProvider>
    );
  };
};

export type FlowMetadata = {
  flow:
    | 'signIn'
    | 'signUp'
    | 'userButton'
    | 'userProfile'
    | 'userVerification'
    | 'organizationProfile'
    | 'createOrganization'
    | 'organizationSwitcher'
    | 'organizationList'
    | 'oneTap'
    | 'blankCaptcha'
    | 'waitlist'
    | 'checkout'
    | 'planDetails'
    | 'pricingTable';
  part?:
    | 'start'
    | 'emailCode'
    | 'phoneCode'
    | 'phoneCode2Fa'
    | 'totp2Fa'
    | 'backupCode2Fa'
    | 'password'
    | 'resetPassword'
    | 'emailLink'
    | 'emailLinkVerify'
    | 'emailLinkStatus'
    | 'alternativeMethods'
    | 'forgotPasswordMethods'
    | 'passwordPwnedMethods'
    | 'havingTrouble'
    | 'ssoCallback'
    | 'popupCallback'
    | 'popover'
    | 'complete'
    | 'accountSwitcher';
};

const [FlowMetadataCtx, useFlowMetadata] = createContextAndHook<FlowMetadata>('FlowMetadata');

export const FlowMetadataProvider = (props: React.PropsWithChildren<FlowMetadata>) => {
  const { flow, part } = props;
  const value = React.useMemo(() => ({ value: props }), [flow, part]);
  return <FlowMetadataCtx.Provider value={value}>{props.children}</FlowMetadataCtx.Provider>;
};

export const withFloatingTree = <T,>(Component: React.ComponentType<T>): React.ComponentType<T> => {
  return (props: T) => {
    const parentId = useFloatingParentNodeId();
    if (parentId == null) {
      return (
        <FloatingTree>
          {/* @ts-expect-error */}
          <Component {...props} />
        </FloatingTree>
      );
    }

    /* @ts-expect-error */
    return <Component {...props} />;
  };
};

export { useFlowMetadata };
