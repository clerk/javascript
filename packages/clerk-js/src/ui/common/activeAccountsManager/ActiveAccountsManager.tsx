import { ActiveSessionResource, SessionResource } from '@clerk/types';
import React from 'react';
import { PoweredByClerk } from 'ui/common';
import { useCoreClerk, useCoreSession, useEnvironment } from 'ui/contexts';
import { useNavigate } from 'ui/hooks';
import { useUserButtonPopupVisibility } from 'ui/userButton/contexts/PopupVisibilityContext';
import { windowNavigate } from 'utils';

import AccountSwitcher from './AccountSwitcher';
import { ActiveAccountButtonSet } from './ActiveAccountButtonSet';
import SignOutAll from './SignOutAll';

interface ActiveAccountsManagerProps {
  sessions: SessionResource[];
  navigateAfterSignOutAll: () => void;
  navigateAfterSwitchSession: () => void;
  navigateAfterSignOutOne?: () => void;
  userProfileUrl: string;
  signInUrl: string;
  showActiveAccountButtons?: boolean;
}

export function ActiveAccountsManager({
  sessions,
  navigateAfterSignOutAll,
  navigateAfterSignOutOne,
  navigateAfterSwitchSession,
  signInUrl,
  userProfileUrl,
  showActiveAccountButtons = true,
}: ActiveAccountsManagerProps): JSX.Element {
  const { setSession, signOut } = useCoreClerk();
  const { id: currentSessionId } = useCoreSession();
  const { authConfig } = useEnvironment();
  const { navigate } = useNavigate();
  const [signoutInProgress, setSignoutInProgress] = React.useState(false);
  const [managementNavigationInProgress, setManagementNavigationInProgress] = React.useState(false);
  const { setPopupVisible } = useUserButtonPopupVisibility();

  const handleSignOutSingle = () => {
    setSignoutInProgress(true);
    if (authConfig.singleSessionMode || sessions.length === 0) {
      handleSignOutAll().catch(() => {
        setSignoutInProgress(false);
      });
      return;
    }

    signOut(navigateAfterSignOutOne, { sessionId: currentSessionId }).catch(() => setSignoutInProgress(false));
  };

  const handleManageAccountClick = () => {
    setManagementNavigationInProgress(true);
    navigate(userProfileUrl).finally(() => {
      setManagementNavigationInProgress(false);
      setPopupVisible(false);
    });
  };

  const handleAddAccountClick = () => {
    windowNavigate(signInUrl);
  };

  const handleAccountClick = (session: SessionResource) => {
    // TODO We only set session for active sessions here.
    // In the future, if the passed session is an expired session, we will allow the
    // user to revive it.
    if (session.status === 'active') {
      setSession(session as ActiveSessionResource, navigateAfterSwitchSession);
    }
  };

  const handleSignOutAll = () => {
    return signOut(navigateAfterSignOutAll);
  };

  const shouldRenderAccountSwitcher = sessions.length || !authConfig.singleSessionMode;
  return (
    <div className='cl-active-accounts-manager'>
      {showActiveAccountButtons && (
        <ActiveAccountButtonSet
          handleManageAccount={handleManageAccountClick}
          handleSignout={handleSignOutSingle}
          isSignoutLoading={signoutInProgress}
          isManagementNavigationLoading={managementNavigationInProgress}
        />
      )}
      {shouldRenderAccountSwitcher && (
        <AccountSwitcher
          sessions={sessions}
          isSingleSession={authConfig.singleSessionMode}
          handleAccountClick={handleAccountClick}
          handleAddAccountClick={handleAddAccountClick}
        />
      )}
      {Boolean(sessions.length) && <SignOutAll handleSignOutAll={handleSignOutAll} />}
      <PoweredByClerk className='cl-powered-by-clerk' />
    </div>
  );
}
