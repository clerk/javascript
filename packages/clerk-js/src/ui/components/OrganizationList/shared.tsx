import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
