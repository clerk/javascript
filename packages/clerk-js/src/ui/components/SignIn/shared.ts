import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
