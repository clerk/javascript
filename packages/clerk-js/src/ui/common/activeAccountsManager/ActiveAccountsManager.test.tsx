import { mocked, renderJSON } from '@clerk/shared/testUtils';
import { EnvironmentResource, SessionResource } from '@clerk/types';
import React from 'react';
import { useEnvironment } from 'ui/contexts';

import { ActiveAccountsManager } from './ActiveAccountsManager';

const session = {
  id: 'deadbeef',
  publicUserData: {
    firstName: 'John',
    lastName: 'Doe',
    profileImageUrl: 'http://test.host/profile.png',
    primaryEmailAddress: 'jdoe@example.com',
  },
} as any as SessionResource;

const sessions = [
  session,
  {
    id: 'cafebabe',
    publicUserData: {
      firstName: 'Carla',
      lastName: 'Coe',
      profileImageUrl: 'http://test.host/profile.png',
      primaryEmailAddress: 'ccoe@example.com',
    },
  },
] as any as SessionResource[];

const mockNavigate = jest.fn();
jest.mock('ui/hooks', () => ({
  useNavigate: jest.fn(() => {
    return {
      navigate: mockNavigate,
    };
  }),
}));

jest.mock('ui/contexts', () => {
  return {
    useEnvironment: jest.fn(),
    useCoreClerk: () => {
      return {
        setSession: jest.fn(),
      };
    },
    useCoreSessionList: () => {
      return sessions;
    },
    useCoreSession: () => {
      return session;
    },
  };
});

const noop = () => {
  //
};

const singleSessionResource = {
  authConfig: {
    singleSessionMode: true,
  },
  displayConfig: {
    branded: false,
  },
} as any as EnvironmentResource;

const multiSessionResource = {
  authConfig: {
    singleSessionMode: false,
  },
  displayConfig: {
    branded: false,
  },
} as any as EnvironmentResource;

describe('ActiveAccountsManager', () => {
  it('renders a switcher with all the available sessions', () => {
    mocked(useEnvironment as jest.Mock<EnvironmentResource>, true).mockImplementation(() => multiSessionResource);

    const tree = renderJSON(
      <ActiveAccountsManager
        sessions={sessions}
        navigateAfterSignOutAll={noop}
        navigateAfterSwitchSession={noop}
        userProfileUrl={'user_profile_url'}
        signInUrl={'sign_in_url'}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('does not render active accounts with only a single session', () => {
    mocked(useEnvironment as jest.Mock<EnvironmentResource>, true).mockImplementation(() => multiSessionResource);

    const tree = renderJSON(
      <ActiveAccountsManager
        showActiveAccountButtons={false}
        sessions={[session]}
        navigateAfterSignOutAll={noop}
        navigateAfterSwitchSession={noop}
        userProfileUrl={'user_profile_url'}
        signInUrl={'sign_in_url'}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders all active sessions even in single session resource mode', () => {
    mocked(useEnvironment as jest.Mock<EnvironmentResource>, true).mockImplementation(() => singleSessionResource);

    const tree = renderJSON(
      <ActiveAccountsManager
        showActiveAccountButtons={false}
        sessions={[session]}
        navigateAfterSignOutAll={noop}
        navigateAfterSwitchSession={noop}
        userProfileUrl={'user_profile_url'}
        signInUrl={'sign_in_url'}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
