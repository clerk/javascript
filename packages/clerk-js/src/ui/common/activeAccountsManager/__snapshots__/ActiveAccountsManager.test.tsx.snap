// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveAccountsManager does not render active accounts with only a single session 1`] = `
<div
  className="cl-active-accounts-manager"
>
  <div
    className="cl-account-switcher"
  >
    <button
      className="button text primary hoverable cl-account-switcher-button"
      onClick={[Function]}
    >
      <img
        alt="John Doe"
        className="avatar cl-user-button-avatar"
        height={36}
        onError={[Function]}
        src="http://test.host/profile.png?height=64"
        title="John Doe"
        width={36}
      />
      <div
        className="cl-details"
      >
        <span>
          John
           
          Doe
        </span>
        <span />
      </div>
    </button>
    <button
      className="button text primary hoverable cl-account-switcher-button cl-account-switcher-add-button"
      onClick={[Function]}
    >
      <div
        className="cl-account-switcher-add-button-icon"
      >
        <svg
          data-filename="../shared/assets/icons/user-plus.svg"
        />
      </div>
      <span>
        Add account
      </span>
    </button>
  </div>
  <div
    className="cl-sign-out-all"
  >
    <button
      className="button outline primary hoverable cl-sign-out-all-button small"
      onClick={[Function]}
    >
      <svg
        data-filename="../shared/assets/icons/sign-out.svg"
      />
      Sign out of all accounts
    </button>
  </div>
</div>
`;

exports[`ActiveAccountsManager renders a switcher with all the available sessions 1`] = `
<div
  className="cl-active-accounts-manager"
>
  <div
    className="cl-active-account-buttonset"
  >
    <button
      className="button outline primary hoverable cl-accounts-manager-button small"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <svg
        data-filename="../shared/assets/icons/cog.svg"
      />
      Manage account
    </button>
    <button
      className="button outline primary hoverable cl-accounts-manager-button small"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <svg
        data-filename="../shared/assets/icons/sign-out.svg"
      />
      Sign out
    </button>
  </div>
  <div
    className="cl-account-switcher"
  >
    <button
      className="button text primary hoverable cl-account-switcher-button"
      onClick={[Function]}
    >
      <img
        alt="John Doe"
        className="avatar cl-user-button-avatar"
        height={36}
        onError={[Function]}
        src="http://test.host/profile.png?height=64"
        title="John Doe"
        width={36}
      />
      <div
        className="cl-details"
      >
        <span>
          John
           
          Doe
        </span>
        <span />
      </div>
    </button>
    <button
      className="button text primary hoverable cl-account-switcher-button"
      onClick={[Function]}
    >
      <img
        alt="Carla Coe"
        className="avatar cl-user-button-avatar"
        height={36}
        onError={[Function]}
        src="http://test.host/profile.png?height=64"
        title="Carla Coe"
        width={36}
      />
      <div
        className="cl-details"
      >
        <span>
          Carla
           
          Coe
        </span>
        <span />
      </div>
    </button>
    <button
      className="button text primary hoverable cl-account-switcher-button cl-account-switcher-add-button"
      onClick={[Function]}
    >
      <div
        className="cl-account-switcher-add-button-icon"
      >
        <svg
          data-filename="../shared/assets/icons/user-plus.svg"
        />
      </div>
      <span>
        Add account
      </span>
    </button>
  </div>
  <div
    className="cl-sign-out-all"
  >
    <button
      className="button outline primary hoverable cl-sign-out-all-button small"
      onClick={[Function]}
    >
      <svg
        data-filename="../shared/assets/icons/sign-out.svg"
      />
      Sign out of all accounts
    </button>
  </div>
</div>
`;

exports[`ActiveAccountsManager renders all active sessions even in single session resource mode 1`] = `
<div
  className="cl-active-accounts-manager"
>
  <div
    className="cl-account-switcher"
  >
    <button
      className="button text primary hoverable cl-account-switcher-button"
      onClick={[Function]}
    >
      <img
        alt="John Doe"
        className="avatar cl-user-button-avatar"
        height={36}
        onError={[Function]}
        src="http://test.host/profile.png?height=64"
        title="John Doe"
        width={36}
      />
      <div
        className="cl-details"
      >
        <span>
          John
           
          Doe
        </span>
        <span />
      </div>
    </button>
  </div>
  <div
    className="cl-sign-out-all"
  >
    <button
      className="button outline primary hoverable cl-sign-out-all-button small"
      onClick={[Function]}
    >
      <svg
        data-filename="../shared/assets/icons/sign-out.svg"
      />
      Sign out of all accounts
    </button>
  </div>
</div>
`;
